### ISSUE TITLE
[FEATURE] Support for zero length CRLF keepalives

### ISSUE THREAD

**Issue by Lt-Flash** [OpenSIPS/opensips#2817]
**Is your feature request related to a problem? Please describe.**
Currently OpenSIPS doesn't allow SIP to process CRLF keepalives coming from multiple soft and desk phones. It just produces an error like this in logs:
```
Apr 22 17:25:15 INFO:core:parse_first_line: message too short: 0
Apr 22 17:25:15 INFO:core:parse_first_line: bad message
Apr 22 17:25:15 ERROR:core:parse_msg: message=<

>
Apr 22 17:25:15 ERROR:core:receive_msg: Unable to parse msg received from [X.X.X.X:34699]
Apr 22 17:25:15 ERROR:proto_wss:ws_process: receive_msg failed
```

**Describe the solution you'd like**
We'd like to have some mechanism to allow processing of those, even a very primitive one - if such keepalive comes from a registered or otherwise known SIP device - send a response to it instead of generating an error message in logs.

**Implementation**
  - Component: Probably UDP/TCP/TLS/WSS module?
  - Type: Function like 'sl_send_reply' or 't_reply' if we decide to answer to such keep alive message
  - Name: Up to devs to decide

**Describe alternatives you've considered**
Nothing I was able to do as currently OpenSIPS doesn't even start a script to process such messages

**Comment by bogdan-iancu** [comment]
@Lt-Flash , have you tried these ?
* https://opensips.org/html/docs/modules/3.2.x/proto_tls.html#param_tls_crlf_pingpong
* https://opensips.org/html/docs/modules/3.2.x/proto_tls.html#param_tls_crlf_drop

**Comment by Lt-Flash** [comment]
Hi Bogdan,
These ones are for TLS and I'm aware of them, but Yealink handsets, for
example send CRLF keepalives even in UDP mode!

Regards,
Yury.

**Comment by bogdan-iancu** [comment]
There are also for TCP
https://opensips.org/html/docs/modules/3.2.x/proto_tcp.html#idp5870208

For UDP, whatever package shorter than 20 bytes is automatically discard by the UDP layer, so no parsing errors should be generated here.

**Comment by Lt-Flash** [comment]
I think in UDP I'm getting something like this in OpenSIPS logs when phone is sending CRLF:
```
[May  3 06:43:28] INFO:core:parse_first_line: empty  or bad first line
[May  3 06:43:28] INFO:core:parse_first_line: bad message
[May  3 06:43:28] ERROR:core:parse_msg: message=<>
[May  3 06:43:28] ERROR:core:receive_msg: Unable to parse msg received from [1XX.XXX.XXX.XXX:5060]
```

**Comment by Lt-Flash** [comment]
Asterisk PBX would just reply to such message while OpenSIPS throws an error without any reply and processing and that stops keepalives from working, that's why I created this feature request. Thanks!

**Comment by bogdan-iancu** [comment]
Again, UDP packages shorter than 20 bytes are discarded. Provide a pcap to show the actual UDP package as sent to OpenSIPS

**Comment by bogdan-iancu** [comment]
I see payload of 4, so they will be discarded by OpenSIPS. Maybe the errors you see are triggered by something else. What opensips version do you have?

**Comment by Lt-Flash** [comment]
This error message is pretty much the same for both of my OpenSIPS builds:
```
version: opensips 3.2.4 (x86_64/linux)
flags: STATS: On, DISABLE_NAGLE, USE_MCAST, SHM_MMAP, PKG_MALLOC, Q_MALLOC, F_MALLOC, HP_MALLOC, DBG_MALLOC, FAST_LOCK-FUTEX-ADAPTIVE_WAIT
ADAPTIVE_WAIT_LOOPS=1024, MAX_RECV_BUFFER_SIZE 262144, MAX_LISTEN 16, MAX_URI_SIZE 1024, BUF_SIZE 65535
poll method support: poll, epoll, sigio_rt, select.
git revision: 50407d340
main.c compiled on 03:02:10 Feb 14 2022 with gcc 9
```

And this one:
```
version: opensips 2.4.8 (x86_64/linux)
flags: STATS: On, DISABLE_NAGLE, USE_MCAST, SHM_MMAP, PKG_MALLOC, F_MALLOC, FAST_LOCK-FUTEX-ADAPTIVE_WAIT
ADAPTIVE_WAIT_LOOPS=1024, MAX_RECV_BUFFER_SIZE 262144, MAX_LISTEN 16, MAX_URI_SIZE 1024, BUF_SIZE 65535
poll method support: poll, epoll, sigio_rt, select.
git revision: 1e891b5e3
main.c compiled on 23:23:48 Sep 30 2020 with gcc 7
```

Both are built from source code from releases, not GIT.

Yealink and other handsets/softphones are using this UDP CRLF to keep connection via NAT router alive, otherwise router would close that NAT entry if it can't see any traffic for 30 seconds (by default for many routers for UDP).

**Comment by bogdan-iancu** [comment]
There is something wrong with the way you test.

A simple test line
```
echo -e '\r\n\r\n' | nc -u 127.0.0.1 5060
```
against an opensips running in debug mode resulted only in:

```
May  3 15:34:07 [54489] DBG:core:udp_read_req: probing packet received len = 5
``` 
no error, nothing else

**Comment by Lt-Flash** [comment]
Yes, I confirm that's correct, must be something else, I will try to do a pcap of what exactly happens when OpenSIPS throws that error, thank you!

**Comment by Sandcrawler** [comment]
Hello,
I do not mean to necro a thread, but I was reading through some of the tickets and believe I have seen this before... and would like to offer something if I can.

Our network was subject to some more than **exceptional** circumstances several months back. We run some much older versions in our production environment with similarities to what has been described in OP issue. I saw in our environment, at a certain point in what we suffered, this exact error. 

There is an RFC and the messaging coming from the handsets appears valid, and from all I can work out it is. Something else was happening that pinned our environment to the floor. 

Where X is the length that feels right... 
-A INPUT -m udp -p udp -m length --length 0:X -j DROP

Reconfigure to use SIP keepalive mechanisms and kick it out.
Good luck soldier.
